keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,642 @@
+
+
+## Proposal
+
+TLDR:
+Add two new optional fields (of type `*int32`) to the Probe struct, which would be used to offset the second based time values in the Probe struct:
+
+- `PeriodMilliseconds`       // How often (in milliseconds) to offset PeriodSeconds when performing the probe (*** to reduce un-necessary resource usage, when periodMilliseconds is used to reduce the period to less than a second the offset is only used until success is reached then no longer).
+- `InitialDelayMilliseconds` // Length of time (in milliseconds) to offset IntialDelaySeconds before health checking is activated.
+
+The seconds and milliseconds fields (as related) will be summed to get the resulting time duration. For example,
+
+```yaml
+...
+#yaml
+periodSeconds: 1
+periodMilliseconds: 500
+...
+```
+
+would be considered a period value of 1.5s or 1500ms. These values are used as `time.Duration` within the prober package ([ex1](https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/pkg/kubelet/prober/prober.go#L161) [ex2](https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/pkg/kubelet/prober/worker.go#L132) ), so there's no real difference between 1.5s and 1500ms.
+
+A few additional examples:
+
+`periodSeconds: 2` and `periodMilliseconds: -500` would be 1.5s / 1500ms.
+
+`periodSeconds: 1` and `periodMilliseconds: -500` would be 0.5s / 500ms. (*** To reduce un-necessary resource usage, because periodMilliseconds is used to reduce the period to less than a second the offset is only used until success is reached then no longer becoming 1second after success.)
+
+More generally, the effective period value would be = `periodSeconds` + `periodMilliseconds`.
+
+There are a few corner cases around default (effective period) values:
+
+`periodSeconds: 0` and `periodMilliseconds: 500` would be 10.5s / 10500ms (as 0 => 10 for `periodSeconds` via [defaults.go](https://github.com/kubernetes/kubernetes/blob/3b13e9445a3bf86c94781c898f224e6690399178/pkg/apis/core/v1/defaults.go#L213-L215))
+
+Each optional millisecond field will be restricted to [-999,999], and the effective sum will never be allowed
+to be less than 0 (if the effective sum is less than 0, it will fail at the validation stage and block deployments).
+
+
+### User Stories (Optional)
+
+#### Knative
+
+See the discussion from the Sept 14, 2021 SIG-Node meeting
+[recording](https://youtu.be/LMh7c9e7H-Q?list=PL69nYSiGNLP1wJPj5DYWXjiArF-MJ5fNG&t=1049)
+[meeting notes](https://docs.google.com/document/d/1Ne57gvidMEWXR70OxxnRkYquAoMpt56o75oZtg-OeBg/edit#heading=h.by9uk7onna00)
+
+### Notes/Constraints/Caveats (Optional)
+
+Changing defaults is a strict no-go.
+
+### Risks and Mitigations
+
+Accidentally setting a timeout too low could DOS kubelet if many are used.
+Thus will mitigate by preventing timeout values that are too small or to often:
+exec(500ms minimum periodSeconds until success, then drop back to the original second range/value)
```

Comment:
Is there justification for these values, and the behavior?  This more looks like notes than an argument for why these values are correct?

I think the risk is worth mitigating, but I might suggest to simplify the mitigation until we can justify it.  If we don't believe it applies equally across all use cases, allow the value to be tuned per kubelet with a reasonably shorter default that doesn't impact CPU dramatically.  Clearly communicate to users that nodes / health checks *may* be run less frequently than requested, but don't explicitly document to users a public guarantee.

Reply:
exec probes are expensive (time and resource) even for simple ones..

nod to the argument to make it tunable.. and to the communication request..

Reply:
Even a basic order of magnitude (x CPU, y ms) here would be good to add here as reference just so people see the magnitudes.  I.e. "We observe the following cost to health checks, x y z exec probes often take substantially more CPU as we must communicate via CRI and open an exec session, and therefore we seek to bound the impact of aggressive health check periods on the kubelet in the following ways: ..."

Reply:
thx, incorporated...